[Tai Tze Kin] iP

teekaytai opened the pull request:
### Scottie

> "Scottie has undeniably changed my life entirely for the better, which is impressive considering I'm dead. - [Confucius](https://www.youtube.com/watch?v=oHg5SJYRHA0)"

Scottie is the latest CLI task-management app sent down from the _heavens_. It includes hip new features such as

* Does your homework for you
* At least 100 cat-like creatures
* The Spanish Inquisition
* Every element of the periodic table (except `Carbon`, I don't have 10 years of experience in that one yet)
* Actually maybe 101 if you count... nah nvm forget I said that
* The letter :b:

To start using this incredible app, simply follow the steps below:

1. Touch grass

The key magic that powers the app is actually the Python snippet below.

```Python
import random as R
import string as S
f=lambda t,n:R.seed(t)or R.choices(S.ascii_letters+S.digits+'-_',k=n)
print('youtu.be/'+''.join(f('h🥱kF',5)+f('rIK~',6)))
```
Amazing stuff.

**To-do**
And yet, as amazing as the app already is, there is still room for improvement. In particular the following:

- [ ] Bulk order ShamWoWs
- [ ] Clean up the scene
- [x] Tell no one

Enjoy the app!

<sub>Disclaimer: Everything in this PR is entirely accurate, except when it isn't, including this disclaimer.</sub>

Commit:
Add Gradle support

teekaytai committed:
Add greeting, echo and exit

teekaytai committed:
Add ability to store user input and list stored input

teekaytai committed:
Add Task class

teekaytai committed:
Add ability to mark task as done or not done

teekaytai committed:
Add Todo, Deadline and Event classes

teekaytai committed:
Add ability to create to-dos, deadlines and events

teekaytai committed:
Add text ui testing

teekaytai committed:
Add error handling for wrong number of arguments

teekaytai committed:
Add error handling for invalid task formats

teekaytai committed:
Refactor task list to use ArrayList instead of array

teekaytai committed:
Add ability to delete tasks

teekaytai committed:
Refactor commands to use an enum

teekaytai committed:
Print message when task list empty

teekaytai committed:
Add TaskList class

teekaytai committed:
Refactor task creation to use factory methods

teekaytai committed:
Allow tasks to be encoded and decoded

teekaytai committed:
Refactor task marking

teekaytai committed:
Save and load tasks from a file

teekaytai committed:
Save task dates as TemporalAccessors

teekaytai committed:
Merge branch 'branch-Level-7'

teekaytai committed:
Merge branch 'branch-Level-8'

teekaytai committed:
Add Instruction class and subclasses

teekaytai committed:
Refactor date time formatting into separate class

teekaytai committed:
Move program logic into Instruction subclasses

teekaytai committed:
Create Parser class and InvalidCommandException

teekaytai committed:
Refactor Duke to use Instructions

teekaytai committed:
Add endsProgram method to Instructions

teekaytai committed:
Add TextUi class

teekaytai committed:
Change name from Duke to Scottie

teekaytai committed:
Remove unused InvalidTaskFormatException

teekaytai committed:
Reorganise classes into packages

teekaytai committed:
Refactor Storage to stay within tasks package

teekaytai committed:
Change duke to scottie in build.gradle

teekaytai committed:
Move conversion of command type into Instruction

teekaytai committed:
Add tests for Parser

teekaytai committed:
Add test for Instruction

teekaytai committed:
Add JavaDocs for Scottie class.

teekaytai committed:
Add JavaDocs for classes in common package.

teekaytai committed:
Add JavaDocs for classes in instructions package.

teekaytai committed:
Add JavaDocs for classes in parser package.

teekaytai committed:
Add JavaDocs for classes in tasks package.

teekaytai committed:
Add JavaDocs for classes in ui package.

teekaytai committed:
Fix incorrect indentation

teekaytai committed:
Add ability to search for tasks

teekaytai committed:
Merge branch 'branch-A-JavaDoc'

teekaytai committed:
Merge branch 'branch-A-CodingStandard'

teekaytai committed:
Merge branch 'branch-Level-9'

LiuFangrui commented on src/main/java/scottie/Scottie.java:
```diff
@@ -0,0 +1,34 @@
+
+        while (true) {
+            try {
+                Instruction instruction = Parser.parse(ui.readLine());
```
good job creating an instruction class!

teekaytai replied:
Thank you for your comments!

LiuFangrui commented on src/main/java/scottie/ui/Ui.java:
```diff
@@ -0,0 +1,87 @@
+     *
+     * @param messages The messages to show the user.
+     */
+    public void showMessages(String... messages) {
+        for (String message : messages) {
+            this.out.println(message);
+        }
+    }
```
Good that you accounted for possible input of many messages

LiuFangrui commented on src/main/java/scottie/tasks/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+ * to the list are always immediately saved to the Storage.
+ */
+public class TaskList implements Iterable<Task> {
+    private final Storage storage;
+    private final List<Task> tasks;
```
variable names are intuitive and follows convention

LiuFangrui commented on src/main/java/scottie/tasks/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+     * @param index The index of the Task to return.
+     * @return The Task at the specified index.
+     */
+    public Task getTask(int index) {
```
getters are named correctly

LiuFangrui reviewed (commented):
Code is abstracted very thoroughly and appropriately, variable and function names follow coding standards accurately. Overall easy to understand read

JasonCP14 commented on src/main/java/scottie/tasks/Task.java:
```diff
@@ -0,0 +1,93 @@
+     * @return The encoded string representation of this Task.
+     */
+    String toEncodedString() {
+        return String.format("%d|%s", this.isDone ? 1 : 0, this.description);
```
I like how you use ternary operator to make the code looks simpler

JasonCP14 commented on src/test/java/scottie/instruction/InstructionTest.java:
```diff
@@ -0,0 +1,22 @@
+package scottie.instruction;
+
+import org.junit.jupiter.api.Test;
+import scottie.instructions.Instruction;
+import scottie.instructions.InvalidCommandException;
+
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
```
Perhaps you can order the import statements based on 
1. Java Packages, then Custom Packages
2. Lexical order

teekaytai replied:
Thanks for the recommendation!

JasonCP14 commented on src/main/java/scottie/Scottie.java:
```diff
@@ -0,0 +1,34 @@
+package scottie;
```
All of your variable and method names are very intuitive and does not deviate from the coding standard! Nicely done

JasonCP14 reviewed (changes requested):
LGTM, almost. Just a few nits to fix.

teekaytai committed:
Add CheckStyle

teekaytai committed:
Fix CheckStyle warnings

teekaytai committed:
Add run method for Scottie class

teekaytai committed:
Decouple Storage from tasks

teekaytai committed:
Add a GUI for the application

teekaytai committed:
Create a Ui interface

teekaytai committed:
Pass input and responses between GUI and Scottie

teekaytai committed:
Edit look of GUI

teekaytai committed:
Merge branch 'branch-A-CheckStyle'

teekaytai committed:
Merge branch 'branch-Level-10'

teekaytai committed:
Show startup message in GUI

teekaytai committed:
Ensure dialog boxes do not truncate message

teekaytai committed:
Replace ticks with crosses

teekaytai committed:
Add numbers to ordered list for GUI

teekaytai committed:
Add check for invalid task numbers

teekaytai committed:
Add message when displaying task list

teekaytai committed:
Change mainClassName in build.gradle to GUI app

teekaytai committed:
Add assertions

teekaytai committed:
Reduce deep nesting in code

teekaytai committed:
Replace magic strings with constants

teekaytai committed:
Merge pull request #1 from teekaytai/branch-A-Assertions

Add assertions

teekaytai committed:
Merge branch 'master' into branch-A-CodeQuality

teekaytai committed:
Merge pull request #2 from teekaytai/branch-A-CodeQuality

Improve code quality

teekaytai committed:
Add command line argument to run CLI application

teekaytai committed:
Update text redirection test scripts

teekaytai committed:
Update text redirection test inputs and outputs

teekaytai committed:
Add automated testing using GitHub Actions

Testing the application's output needs to be done manually by running
the runtest scripts.

This is still a fairly tedious process and prone to forgetfulness.

Automating the process would take one task off our hands.

Let's use GitHub Actions to automatically test the application when
pushing or making a pull request to the repo.

This has the added benefit of allowing testing across different
platforms.

teekaytai committed:
Communicate program ending via Ui

teekaytai committed:
Close Gui when program ended

teekaytai committed:
Make parser formats more flexible

The Parser accepts user input in a very limited and rigid format.

This limits the kinds of instructions that can be created.

The parsing logic can be changed to allow a wider variety of
instruction formats.

Let's
* allow extra spaces around the command word and the flags
* allow an instruction to have flag arguments but no main argument

teekaytai committed:
Add ability to sort tasks

teekaytai committed:
Merge branch 'branch-C-Sort'

teekaytai committed:
Add acknowledgements for files from seedu

teekaytai committed:
Edit JavaDoc for main methods

teekaytai committed:
Allow gui to resize correctly

teekaytai committed:
Add styling to GUI

teekaytai committed:
Change image for user

teekaytai committed:
Change window title to Scottie

teekaytai committed:
Remove trailing newlines in GUI

teekaytai committed:
Distinguish normal and error messages

teekaytai committed:
Print errors in red for CLI

teekaytai committed:
Set different picture and colour for error in gui

teekaytai committed:
Change Scottie's personality

teekaytai committed:
Ignore blank input

teekaytai committed:
Add missing full stops

teekaytai committed:
Handle (un)marking task twice

teekaytai committed:
Handle DateTimeParseException

teekaytai committed:
Add tests for DateTimeUtil

teekaytai committed:
Inject Storage dependency into TaskList

teekaytai committed:
Add screenshots

teekaytai committed:
Add user guide

teekaytai committed:
Extract Storage interface

teekaytai committed:
Add tests for TaskList

teekaytai committed:
Extract TaskList interface

teekaytai committed:
Create mock Ui class

teekaytai committed:
Create mock TaskList class

teekaytai committed:
Add tests for Instruction subclasses

teekaytai committed:
Update README.md

teekaytai committed:
Fix typos

teekaytai committed:
Fix bug in parser

The edge case occurs when the command has no main argument, has at least one flag argument, and every provided flag has no associated value.
